### ISSUE TITLE
Issues when used inside LeBlender grid editor

### ISSUE THREAD

**Issue by bjarnef** [rasmusjp/umbraco-multi-url-picker#63]
I have tried to use the multi url picker as a `link` property in a LeBlender grid editor.
However I can't really make it work with cast to the strongly typed object.

I have tried with both a Multi Url Picker and Single Url Picker (max items = 1).

Inside LeBlender grid editor.

```
@if (Model.Items != null && Model.Items.Any())
{
     var item = Model.Items.First();
     
     // both of these return null
     var link1 = item.GetValue<IEnumerable<Link>>("link");
     var link2 = item.GetValue<Link>("link");
}
```

If I use LeBlenders `GetRawValue` method I get the raw json, but when I deserialize this to either `IEnumerable<Link>` or `Link` it returns `Unexpected character encountered while parsing value: l. Path '', line 0, position 0.`
```
var link = item.GetRawValue("link") 
var linkPage = !string.IsNullOrWhiteSpace(link) ? JsonConvert.DeserializeObject<Link>("link") : null;
```

**Comment by bjarnef** [comment]
Should it always return an json object with array of objects instead of a json array?

**Comment by bjarnef** [comment]
I wonder if it is because the `Link` model only has getters.

If I create my own model class like the following it does work.

```
public class LinkModel
{
    public int? Id { get; set; }
    public string Name { get; set; }
    public string Target { get; set; }
    public LinkType Type { get; set; }
    public Udi Udi { get; set; }
    public string Url { get; set; }
}
```

```
JArray jArray = JsonConvert.DeserializeObject<JArray>(link);
var jsonObj = jArray.First.ToString();

var linkObject = JsonConvert.DeserializeObject<LinkModel>(jsonObj);
if (linkObject != null)
{
    @linkObject.Id<br />
    @linkObject.Udi<br />
    @linkObject.Name<br />
    @(linkObject.Target ?? "_self")<br />
    @linkObject.Type<br />
    @linkObject.Url<br />
}
```

![image](https://user-images.githubusercontent.com/2919859/32373279-cfd69e0a-c098-11e7-9e9d-b6b78e4477c5.png)

**Comment by bjarnef** [comment]
This is the fix I ended up with for now to make it work inside LeBlender grid editor as a Single Url Picker (max items = 1).

```
var item = Model.Items.First();
var link = item.GetRawValue("link");

// A temp fix to get Single Url Picker value (might need a few changed for Multi Url Picker).
JToken jToken = JsonConvert.DeserializeObject<JToken>(link);
Link linkPage = jToken != null && jToken.First != null ? new Link(jToken.First) : null;
```

**Comment by bjarnef** [comment]
It seems to be an issue in Nested Content as well.

![image](https://user-images.githubusercontent.com/2919859/32481786-340dc5b6-c395-11e7-9547-8112b67d4da9.png)

```
@{
    var advertisers = Model.Content.GetPropertyValue<IEnumerable<IPublishedContent>>("advertisements") ?? Enumerable.Empty<IPublishedContent>();
}

@if (advertisers.Any())
{
    <div class="section row suppliers border-top">
        <div class="col-xs-10">
            <span class="headline">Annoncører</span>
        </div>
        @foreach (var banner in advertisers.OfType<Advertisement>())
        {
            DateTime startTime = banner.StartTime;
            DateTime endTime = banner.EndTime;

            var image = banner.Image;
            var link = banner.Link;
            var summaryLink = banner.ProductSummary;

            if (DateTime.Now > startTime && DateTime.Now < endTime && image != null)
            {
                <div class="col-sm-2">
                    ...
                </div>
            }
        }
    </div>
}
```

**Comment by bjarnef** [comment]
Here it has a null or empty check.
https://github.com/rasmusjp/umbraco-multi-url-picker/blob/master/src/RJP.MultiUrlPicker/MultiUrlPickerValueConverter.cs#L28

But only a null check here.
https://github.com/rasmusjp/umbraco-multi-url-picker/blob/master/src/RJP.MultiUrlPicker/MultiUrlPickerValueConverter.cs#L38-L43

**Comment by rasmusjp** [comment]
Thanks for the very detailed report(s).

I'm not familiar with LeBlender, but I've just tried setting it up on a clean Umbraoc 7.7.4 and LeBlender 1.0.8.2 and I cannot reproduce the problem using your code samples.

For a new data type configured with max items 1, `@(item.GetValue<Link>("link"))` correctly returns the link, and the same with multiple links. So I'm not sure why it's not working for you.


Regarding the exception when switching from LinkPicker to MultiUrlPicker is because I always store the data as an array and by looking at the source for LinkPicker it's stored as an object.
When you publish the node, the data that's stored is removed and replaced with an array which the value converter expects.


The string null or empty and null checks should be fine since Umbraco always calls `ConvertDataToSource` and passes the return data to `ConvertSourceToObject`, see https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Web/PublishedCache/XmlPublishedCache/XmlPublishedProperty.cs#L50 for reference

**Comment by bjarnef** [comment]
Okay, I am using LeBlender 1.0.8.4 and Umbraco 7.6.11

Do you have the core property value converters enabled?

I have noticed some issues with some property editors inside LeBlender, when either using @Jeavon 's core property value converter or the core property value converters included in Umbraco 7.6+

https://github.com/Jeavon/Umbraco-Core-Property-Value-Converters/issues/24

The issue regarding switching from LinkPicker to MultiUrlPicker might have been because the property value contained a whitespace or `JArrray` returned null and it here use `.ToString()`
https://github.com/rasmusjp/umbraco-multi-url-picker/blob/master/src/RJP.MultiUrlPicker/MultiUrlPickerValueConverter.cs#L43

https://github.com/rasmusjp/umbraco-multi-url-picker/blob/3579a38c4465dfd35f90456daea43c06c3fea64b/src/RJP.MultiUrlPicker/Models/MultiUrls.cs#L22-L27

**Comment by rasmusjp** [comment]
Yes, the value converters were enabled. I'll try to see if I can reproduce it with 7.6.11

I'll add some error handling if the data stored in the property is not what's expected.

**Comment by rasmusjp** [comment]
Hmm.. I still can't reproduce this with Umbraco 7.6.11, but as @Jeavon pointed out they are "sometimes" returning the source value which they shouldn't.

Also I've noticed they don't call any of the `IValueEditor` methods (e.g. `ConvertEditorToDb` or `ConvertDbToEditor`) when retrieving or saving the data, which could cause the data to be stored wrong.

So I'm pretty sure this is a problem with LeBlender and not MultiUrlPicker

**Comment by bjarnef** [comment]
Hi @rasmusjp 

Yes, there might be some issue with LeBlender as @Jeavon mentions here https://github.com/Lecoati/LeBlender/issues/80

However another similar issue I have, where Multi Url Picker not is used inside LeBlender editor but inside Nested Content. When deploying from local to live on Umbraco Cloud.
I do have a Nested Content inside LeBlender, which seem to fail when deploying:
https://github.com/umbraco/Umbraco.Deploy.Contrib/pull/10

Because of this it fails on frontpage on live enviroment in another partial, which use Multi Url Picker inside Nested Content.
![image](https://user-images.githubusercontent.com/2919859/32593946-52282046-c52a-11e7-9fa8-eeeb0678dde5.png)

It seems to fails in `ConvertDataToSource`. It did seems to work with the property Multi Url Picker inside Nested Content before adding the other LeBlender editor with Nested Content on the page. So maybe because of the deploy failed, the Multi Url Picker has empty or not valid json?

**Comment by bjarnef** [comment]
Okay, for the last issue I mentioned with Multi Url Picker inside Nested Content it seems to be fixed after upgrading Multi Url Picker to v2.1.0

Now I don't get the ysod although it fails deploying e.g. frontpage with a LeBlender grideditor with Nested Content.

**Comment by bjarnef** [comment]
However I still have the issue with the cast inside LeBlender.

```
JToken jToken = JsonConvert.DeserializeObject<JToken>(link);
Link linkPage = jToken != null && jToken.First != null ? new Link(jToken.First) : null; // works
```
```
Link linkPage = item.GetValue<Link>("link"); // doesn't work
```

I am using Umbraco v7.6.11 and LeBlender v1.0.8.4, but it might be because of this issue:
https://github.com/Lecoati/LeBlender/issues/80
